### PR TITLE
feat: add /cid/:cid to linkdex CARs by cid

### PR DESCRIPTION
**Pull request by olizilla**
Try and find a set of CARs that contain a complete DAG for a given CID.

For dotStorage CARs are grouped by userid with keys like `raw/<cid>/<userid>/<hash>.car`.

We find all the CARs for a given cid by prefix, group them by userid, and then sort by count of CARs, under the assumptions that the upload with the most CARs is likely to contain the full DAG.

This is needed for our cron pin status job in dotStorage so we can land https://github.com/web3-storage/web3.storage/pull/1785

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

### CODE REVIEW

**Comment by olizilla** [comment]
This in hold as we don't currently need it. We have the `key` in the dotstroage worker, and after the last s3 upload of a set of partials, the final call as access to all the partial cars. We don't need a cron job to find it later, where we might only have the CID.

This may change if we move to a different s3 key structure.

File: services/functions/linkdex-cid.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * is likely to contain the full DAG.
+ */
+export async function getLinkdexReportForCid (cid: string, bucket: string, s3: S3Client) {
+  if (!cid) {
+    return response({ error: { details: 'cid path param is required' }}, { statusCode: 400 })
+  }
+  if (!CID.parse(cid)) {
+    return response({ error: { details: 'cid is invalid' }}, { statusCode: 400 })
+  }
+  const prefix = `raw/${cid}/`
```
Comment (alanshaw) [inline services/functions/linkdex-cid.ts]: Not blocking, but we should consider CARs in `complete/${cid}`. For examples, CARs from [backup](https://github.com/web3-storage/backup/blob/main/index.js#L133) are written there.

File: services/functions/linkdex-cid.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  for (const carKeys of carGroups) {
+    // create a per user id prefix index. 
+    // avoid using a shared indexer across all here. 
+    // if any user uploaded a CAR with extraneous blocks with links a global index would always return 'Partial'
+    const index = new LinkIndexer()
+    const indexer = indexCar.bind(null, bucket, s3, index)
+    await pMap(carKeys, indexer, { concurrency: 10 })
+    if (index.getDagStructureLabel() === 'Complete') {
+      // return as soon as we find a set of CARs that has a complete DAG
+      return response({ structure: 'Complete' })
```
Comment (alanshaw) [inline services/functions/linkdex-cid.ts]: Shouldn't we return the full report here?

**Review by alanshaw — CHANGES_REQUESTED** [review]

**Comment by alanshaw** [comment]
https://github.com/web3-storage/linkdex-api/pull/11 addresses the two review comments so provided we can merge it here we should be good to go.

**Review by alanshaw — APPROVED** [review]

**Pull request merged**